README.md
# Easy Apache
Easy Apache is a simple script that helps you setup an Apache server, configure it completely, and it also helps you in installing SSL certificates for all your Apache sites. It'll ask for a few inputs, like domain name, do you want to install SSL, etc that's it.

It takes care of configuring Apache, setting up virtual hosts (vhosts), managing different directories for each site, enabling site, disabling sites, and even installing an SSL certificate to your Apache sites.

## Install
Installing easy-apache on to your server is easy too. Just copy & paste below command(s) to your server terminal
##### Using apt
```
sudo add-apt-repository ppa:realpvn/easy-apache
sudo apt update
sudo apt install easy-apache
```
## Run
One command to setup a domain
```
easy-apache [domain]
```
  
## Options & Usage
```
Usage:
easy-apache [options] [domain]

Options:
-s | --ssl          Setup domain with SSL certificate
-h | --help         Help (shows available commands)
-v | --version      Check easy-apache version

Example:
easy-apache -s example.com      # setup domain with ssl certificates
easy-apache example.com         # setup domain (without ssl)
easy-apache -h                  # show help
```

## Works on
- Ubuntu 20.04 (focal)  
(will update more once I test)


debian/changelog
easy-apache (3.1-1) focal; urgency=medium

  * Fixed small issue where banner would not print

 -- Pavan Kumar <realpvn@gmail.com>  Sun, 07 Mar 2021 17:37:44 +0530

easy-apache (3.0-1) focal; urgency=medium

  * Major improvements to how the script runs and handles errors
  * Easier way to add new sites
  * Option to add SSL certificates to already existing sites
  * When domain setup fails, everything done will be reverted, helping keep 
    system clean
  * Fixed version showing not showing as expected
  * Script should be faster than older version

 -- Pavan Kumar <realpvn@gmail.com>  Sun, 07 Mar 2021 16:59:07 +0530

easy-apache (2.13-1) focal; urgency=medium

  * Added man page for more details about easy-apache
  * Improvements while showing banner
  * Removed code which tried to install packages again 

 -- Pavan Kumar <realpvn@gmail.com>  Fri, 16 Oct 2020 21:35:20 +0530

easy-apache (2.12-1) focal; urgency=medium

  * Remove dpkg usage since it will create problems for snap packages 

 -- Pavan Kumar <realpvn@gmail.com>  Wed, 14 Oct 2020 17:46:17 +0530

easy-apache (2.11-1) focal; urgency=medium

  * Remove unneccessary files from user bin

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 18:20:19 +0530

easy-apache (2.10-1) focal; urgency=medium

  * Multiple changes to banner and how it is fetched 

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 14:28:55 +0530

easy-apache (2.9-1) focal; urgency=medium

  * Remove makefile for now as its not neccessary 

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 13:35:48 +0530

easy-apache (2.8-1) focal; urgency=medium

  * Fixed few makefile errors

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 13:19:28 +0530

easy-apache (2.7-1) focal; urgency=medium

  * Major improvements 

 -- Pavan Kumar <realpvn@gmail.com>  Sat, 26 Sep 2020 12:58:41 +0530

easy-apache (2.6-1) focal; urgency=medium

  * Fixed a build error 

 -- Pavan Kumar <realpvn@gmail.com>  Wed, 09 Sep 2020 10:22:52 +0530

easy-apache (2.5-1) focal; urgency=medium

  * Improvements to the way script runs
  * Added better message when SSL installation fails
  * Added a manpage

 -- R Pavan Kumar <realpvn@gmail.com>  Wed, 09 Sep 2020 10:08:42 +0530

easy-apache (2.4-1) focal; urgency=medium

  * Changes to Documentation

 -- R Pavan Kumar <realpvn@gmail.com>  Tue, 08 Sep 2020 08:39:10 +0530

easy-apache (2.3-1) focal; urgency=medium

  * Use certbot from apt repository instead of ppa

 -- R Pavan Kumar <realpvn@gmail.com>  Tue, 08 Sep 2020 07:39:54 +0530

easy-apache (2.2-1) focal; urgency=medium

  * Remove unwanted files

 -- R Pavan Kumar <realpvn@gmail.com>  Mon, 07 Sep 2020 15:30:40 +0530

easy-apache (2.1-1) focal; urgency=medium

  * Fixed server & ssl script missing
  * Combined multiple scripts into one, so execution is easier & less files to worry about

 -- R Pavan Kumar <realpvn@gmail.com>  Mon, 07 Sep 2020 15:03:29 +0530

easy-apache (2.0-1) focal; urgency=medium

  * Initial release

 -- R Pavan Kumar <realpvn@gmail.com>  Sun, 06 Sep 2020 14:03:29 +0530


build.sh
curVer=0
fullCurVer=0
updateDebVersion() {
    oldVer=`head -1 debian/changelog | cut -d '(' -f2 | cut -d ')' -f1`
    echo Old Version: $oldVer
    read -p "New Version: " ver
    dch -v $ver
}

currentVersion() {
    fullCurVer=`head -1 debian/changelog | cut -d '(' -f2 | cut -d ')' -f1`
    curVer=`head -1 debian/changelog | cut -d '(' -f2 | cut -d ')' -f1 | cut -d '-' -f1`
}

makeDeb () {
    echo "Running make"
    clean
    cp easy-apache.sh easy-apache
    read -p "Have you updated version? (Y/N): " verUpdate
    case $verUpdate in
        [Yy]* ) echo "Continuing....";;
        [Nn]* ) read -p "Do you want to update version now? (Y/N): " verUpdate;
                case $verUpdate in
                    [Yy]* ) updateDebVersion;;
                    [Nn]* ) echo "You have to update version to continue";
                            exit;;
                        * ) echo "Invalid input";
                            exit;;
                esac;;
            * ) exit;;
    esac
    currentVersion
    dh_make -p easy-apache_$curVer --indep --createorig -c gpl3 -e realpvn@gmail.com
}

buildDebSource () {
    makeDeb
    debuild -S
    read -p "Do you want to upload Package to PPA? (Y/N): " wantUpload
    case $wantUpload in
        [Yy]* ) uploadPPA;;
    esac
}

buildDebBinary () {
    echo "[!] Binary file cannot to uploaded to PPA"
    read -p "This will deleted any deb source files, do you want to proceed? (Y/N): " cleanNow
    case $cleanNow in
        [Yy]* ) echo "Continuing....";;
        [Nn]* ) echo "You must clean to build binary, exiting..."
                exit;;
            * ) exit;;
    esac
    makeDeb
    debuild
}

debug() {
    if [[ -e ../easy-apache_$fullCurVer.dsc ]]
    then
        echo "Debugging .dsc file"
        lintian ../easy-apache_$fullCurVer.dsc
        echo "----------------------"
        lintian ../easy-apache_${fullCurVer}_all.deb
    else
        echo "You need to build before debugging"
    fi
}

clean () {
    echo "Cleaning build"
    rm -f easy-apache ../easy-apache_*
}

uploadPPA () {
    echo "Current version: $fullCurVer"
    read -p "Is the current version correct? (Y/N): " verCheck
    case $verCheck in
        [Yy]* ) echo "Starting Upload"
                cd ..
                dput ppa:realpvn/easy-apache easy-apache_${fullCurVer}_source.changes;;
        [Nn]* ) echo "Terminating upload";;
            * ) echo "Wrong input";;
    esac
}

if [ "$1" != "" ]
then
    currentVersion
    PARAM=$1
    case $PARAM in
        -ds | --debSource ) buildDebSource;
                            exit;;
        -db | --debBinary ) buildDebBinary;
                            exit;;
        -c  | --clean     ) clean;
                            exit;;
        -d  | --debug     ) debug;
                            exit;;
        -u | --uploadPPA  ) uploadPPA;
                            exit;;
                        * ) echo -e "Usage:\n-ds | --debSource:\t Generates debian source file\n-db | --debBinary:\t Generates debian binary file\n-sn | --snap:\t\t Generates Snap Package\n-c | --clean:\t\t Cleans directory, removes all build files etc\n-u | --uploadPPA:\t Uploads source file to PPA";
                            exit;;
    esac
fi
echo -e "Usage:\n-ds | --debSource:\t Generates debian source file\n-db | --debBinary:\t Generates debian binary file\n-sn | --snap:\t Generates Snap Package\n-c | --clean:\t\t Cleans directory, removes all build files etc\n-u | --uploadPPA:\t Uploads source file to PPA";

setup.sh
# Colors from - https://gist.github.com/5682077.git
TC='\e['

CLR_LINE_START="${TC}1K"
CLR_LINE_END="${TC}K"
CLR_LINE="${TC}2K"

# Hope no terminal is greater than 1k columns
RESET_LINE="${CLR_LINE}${TC}1000D"

Bold="${TC}1m"    # Bold text only, keep colors
Undr="${TC}4m"    # Underline text only, keep colors
Inv="${TC}7m"     # Inverse: swap background and foreground colors
Reg="${TC}22;24m" # Regular text only, keep colors
RegF="${TC}39m"   # Regular foreground coloring
RegB="${TC}49m"   # Regular background coloring
Rst="${TC}0m"     # Reset all coloring and style

# Basic            High Intensity      Background           High Intensity Background
Black="${TC}30m";  IBlack="${TC}90m";  OnBlack="${TC}40m";  OnIBlack="${TC}100m";
Red="${TC}31m";    IRed="${TC}91m";    OnRed="${TC}41m";    OnIRed="${TC}101m";
Green="${TC}32m";  IGreen="${TC}92m";  OnGreen="${TC}42m";  OnIGreen="${TC}102m";
Yellow="${TC}33m"; IYellow="${TC}93m"; OnYellow="${TC}43m"; OnIYellow="${TC}103m";
Blue="${TC}34m";   IBlue="${TC}94m";   OnBlue="${TC}44m";   OnIBlue="${TC}104m";
Purple="${TC}35m"; IPurple="${TC}95m"; OnPurple="${TC}45m"; OnIPurple="${TC}105m";
Cyan="${TC}36m";   ICyan="${TC}96m";   OnCyan="${TC}46m";   OnICyan="${TC}106m";
White="${TC}37m";  IWhite="${TC}97m";  OnWhite="${TC}47m";  OnIWhite="${TC}107m";

#********************************************************************************************

clear
echo "Apache Server Setup"

echo "Updating Server"
sudo apt update && sudo apt upgrade -y
echo "Server updated"

echo "Cleaning after upgrade"
sudo apt autoremove -y && sudo apt autoclean -y

IP=`curl -s icanhazip.com`
echo -e "Server Public IP: ${Purple}"${IP}${Rst}

echo "Installing Apache 2"
sudo apt install apache2 -y
echo -e ${Purple}Apache `apache2 -v`"${Rst}"

echo "Checking UFW"
if sudo ufw status | grep -q inactive$; then
    echo -e "${Red}UFW is disabled. You need to enabled it to continue...${Rst}"
    
    # loop until the answer is yes(Yy) or no(Nn)
    while true
    do
        read -p "Do you want to enable (Yy/Nn)? " yn
        case $yn in
            [Yy]* ) sudo ufw enable;
                    echo "UFW enabled. Allowing SSH & Apache"
                    sudo ufw allow ssh;
                    sudo ufw allow Apache;
                    break;;
            [Nn]* ) echo -e "${Red}Setup Failed to complete${Rst}"; exit;;
                * ) echo "Please answer yes(Yy) or no(Nn) ";;
        esac
    done
else
    echo "UFW enabled"
fi

#TODO(pavank): try to check if site is available at ${IP}
read -p "Number of sites to setup (number only)? " numb
temp=0
while [ $temp != $numb ]
do
    echo -e "${Green}===>${Rst} Setting up site `expr $temp + 1`"
    read -p "Site URL (do not add www, eg input - helloworld.com): " readSiteURL

    #storing urls so it can be used when adding SSL certificate
    siteURL[$temp]=$readSiteURL

    #used for directory name (which is without domain TLD, example.com site folder would be "example" not "example.com")
    siteNameNoTLD=`echo $siteURL[$temp] | cut -d'.' -f1`

    sudo mkdir -p /var/www/$siteNameNoTLD
    sudo chown -R $USER:$USER /var/www/$siteNameNoTLD
    sudo chmod -R 755 /var/www

    #create fake page for temporary viewing
    sudo echo "<h1>Server "`expr $temp + 1`" setup by <a href='https://github.com/realpvn/EasyApacheServer.git'>EasyApacheSetup</a> (https://github.com/realpvn/EasyApacheServer.git) </h1>" > /var/www/$siteNameNoTLD/index.html

    echo "Site ${siteURL[$temp]} created, configuring"
    read -p "Email (leave blank if not required):" siteEmail
    if [ -z $siteEmail ]; then
        siteEmail=dev@localhost
    fi
    echo -e "<VirtualHost *:80>\n\tServerAdmin $siteEmail\n\tServerName ${siteURL[$temp]}\n\tServerAlias www.${siteURL[$temp]}\n\tDocumentRoot /var/www/$siteNameNoTLD\n\tErrorLog \${APACHE_LOG_DIR}/error.log\n\tCustomLog \${APACHE_LOG_DIR}/access.log combined\n</VirtualHost>" | sudo tee /etc/apache2/sites-available/${siteURL[$temp]}.conf

    echo "Enabling site configuration"
    sudo a2ensite ${siteURL[$temp]}.conf
    temp=`expr $temp + 1`
done

echo "Disabling default site (/var/www/html)"
sudo a2dissite 000-default.conf

read -p "Do you also want to install SSL/TSL certificate (Yy/Nn)?" sslReq
case $sslReq in
    [Yy]* ) sudo add-apt-repository ppa:certbot/certbot
            sudo apt install python-certbot-apache -y
            
            while true
            do
                siteCount=0
                siteNameArr=""
                for filePath in /etc/apache2/sites-available/*
                do
                    # how below 'cut' command works
                    # $filePath will have /etc/apache2/sites-available/example.com.conf
                    # first cut will seperate $filePath by '/' and we take everything after field 5 (-f5-) i.e example.com.conf
                    # then we cut example.com.conf by '.' and take everything upto field 2 (-f-2) i.e example.com
                    siteName=`echo $filePath | cut -d'/' -f5- | cut -d'.' -f-2`
                    if [ -e /etc/apache2/sites-available/${siteName}.conf ] && [ ! -e /etc/apache2/sites-available/${siteName}-le-ssl.conf ]
                    then
                        siteNameArr[$siteCount]=$siteName
                        siteCount=`expr $siteCount + 1`
                        echo -e ${Purple}${siteCount}". "$siteName ${Rst}
                    fi
                done
                
                if [ ! $siteCount > 0 ]
                then
                    echo "${Red}SSL Installation exiting. You do not have sites without SSL or it was not found by the script"
                    echo "Please check if your apache server was setup properly and run SSL Installation again${Rst}"
                    break
                fi

                echo -e "${Purple}99. Exit${Rst}"
                read -p "Select site to apply SSL (eg, to exit: 99):" sslSiteSelect
                if [ $sslSiteSelect == 99 ]
                then
                    break
                fi

                #because index starts from 0, but user inputs 1 for 0 hence we subtract 1 by user input
                sslSiteSelect=`expr $sslSiteSelect-1`
                if [ -e /etc/apache2/sites-available/${siteNameArr[$sslSiteSelect]}.conf ] && [ ! -e /etc/apache2/sites-available/${siteURL[$sslSiteSelect]}-le-ssl.conf ]
                then
                    echo "Allowing 'Apache Full' in ufw"
                    sudo ufw delete allow 'Apache'
                    sudo ufw allow 'Apache Full'
                    sudo certbot --apache -d www.${siteURL[$sslSiteSelect]} -d ${siteURL[$sslSiteSelect]}
                    if [ -e /etc/apache2/sites-available/${siteURL[$sslSiteSelect]}-le-ssl.conf ]
                    then
                        echo -e "${Bold}${Green}SSL Successful for ${siteURL[$sslSiteSelect]}${Rst}"
                        break
                    fi
                    echo -e "${Bold}${Red}SSL unSuccessful for ${siteURL[$sslSiteSelect]}${Rst}"
                    break
                fi
                echo -e "${Bold}${Green}SSL already installed for ${siteURL[$sslSiteSelect]}${Rst}"
            done;;
    [Nn]* ) echo "SSL installation rejected by user";;
    * ) echo "Invalid. Rerun the script to add SSL";;
esac

echo "Restarting Apache2 to activate new configuration"
sudo systemctl restart apache2

echo -e "${Bold}${Green}Success! Your sites have been added successfully.${Rst}"
echo "Point your domains A record to $IP and after DNS propagation everything should be working fine."
echo "Sites added and configured are:"

temp=0
while [ $temp != $numb ]
do
    if [ -e /etc/apache2/sites-available/${siteURL[$temp]}-le-ssl.conf ]
    then
        echo "https://${siteURL[$temp]}"
        temp=`expr $temp + 1`
        continue
    fi
    echo "http://"${siteURL[$temp]}
    temp=`expr $temp + 1`
done


rainbow.sh
# Colors from - https://gist.github.com/5682077.git
TC='\e['

CLR_LINE_START="${TC}1K"
CLR_LINE_END="${TC}K"
CLR_LINE="${TC}2K"

# Hope no terminal is greater than 1k columns
RESET_LINE="${CLR_LINE}${TC}1000D"

Bold="${TC}1m"    # Bold text only, keep colors
Undr="${TC}4m"    # Underline text only, keep colors
Inv="${TC}7m"     # Inverse: swap background and foreground colors
Reg="${TC}22;24m" # Regular text only, keep colors
RegF="${TC}39m"   # Regular foreground coloring
RegB="${TC}49m"   # Regular background coloring
Rst="${TC}0m"     # Reset all coloring and style

# Basic            High Intensity      Background           High Intensity Background
Black="${TC}30m";  IBlack="${TC}90m";  OnBlack="${TC}40m";  OnIBlack="${TC}100m";
Red="${TC}31m";    IRed="${TC}91m";    OnRed="${TC}41m";    OnIRed="${TC}101m";
Green="${TC}32m";  IGreen="${TC}92m";  OnGreen="${TC}42m";  OnIGreen="${TC}102m";
Yellow="${TC}33m"; IYellow="${TC}93m"; OnYellow="${TC}43m"; OnIYellow="${TC}103m";
Blue="${TC}34m";   IBlue="${TC}94m";   OnBlue="${TC}44m";   OnIBlue="${TC}104m";
Purple="${TC}35m"; IPurple="${TC}95m"; OnPurple="${TC}45m"; OnIPurple="${TC}105m";
Cyan="${TC}36m";   ICyan="${TC}96m";   OnCyan="${TC}46m";   OnICyan="${TC}106m";
White="${TC}37m";  IWhite="${TC}97m";  OnWhite="${TC}47m";  OnIWhite="${TC}107m";

easyapache.sh
source rainbow.sh

scriptDir="./scripts"
chmod +x $scriptDir/server.sh
chmod +x $scriptDir/ssl.sh

while getopts 'fash' flag; do
  case ${flag} in
    f ) $scriptDir/server.sh
        $scriptDir/ssl.sh
        exit;;
	a ) $scriptDir/server.sh
		exit;;
	s ) $scriptDir/ssl.sh
		exit;;
	h ) echo -e "Usage: easy-apache [options]\nOptions:\n-f:\tFull setup, default option if none is provided\n-a:\tAdding new site (includes apache install)\n-s:\tInstall SSL certificate for sites available\n-h:\tHelp (shows available commands)${Rst}"
       	echo -e "\nExample\n./easyapache -f   #for full installation i.e Apache & SSL certificate\n./easyapache -as  #for installating Apache server & SSL certificate"
		exit;;
    * ) echo -e "Usage: easy-apache [options]\nOptions:\n-f:\tFull setup, default option if none is provided\n-a:\tAdding new site (includes apache install)\n-s:\tInstall SSL certificate for sites available\n-h:\tHelp (shows available commands)${Rst}"
       	echo -e "\nExample\n./easyapache -f   #for full installation i.e Apache & SSL certificate\n./easyapache -as  #for installating Apache server & SSL certificate"
		exit;;
  esac
done
echo -e "Usage: easy-apache [options]\n-f:\tFull setup, default option if none is provided\n-a:\tAdding new site (includes apache install)\n-s:\tInstall SSL certificate for sites available\n-h:\tHelp (shows available commands)${Rst}"
echo -e "\nExample\n./easyapache -f   #for full installation i.e Apache & SSL certificate\n./easyapache -as  #for installating Apache server & SSL certificate"

scripts/ssl.sh
source rainbow.sh
echo -e "${Bold}${Green}Installing SSL${Rst}"
allSitesURL=""
allSitesCount=-1

dpkg -s certbot &> /dev/null
if [ $? -eq 1 ]; then
    echo "Installing Certbot"
    sudo add-apt-repository ppa:certbot/certbot
    sudo apt install python-certbot-apache -y
fi

while true
do
    echo "99 to exit"
    read -p "URL (do not add www, eg input - helloworld.com): " siteName

    if [ $siteName == 99 ]
    then
        echo "Good byee..."
        if [ $allSitesCount == -1 ]
        then
            exit
        fi
        break
    fi

    if [ ! -e /etc/apache2/sites-available/${siteName}.conf ]
    then
        echo -e "${siteName} does not exist.\n${Purple}Add it using './easy-apache -a'${Rst}"
        continue
    fi

    if [ -e /etc/apache2/sites-available/${siteName}-le-ssl.conf ]
    then
        echo -e "${Bold}${Green}${siteName} already has SSL installed${Rst}"
        continue
    fi

    allSitesCount=`expr $allSitesCount + 1`
    allSitesURL[$allSitesCount]=$siteName
    echo -e ${Purple}`expr $allSitesCount + 1`". "$siteName ${Rst}
    
    sudo certbot --apache -d www.$siteName -d $siteName
    if [ -e /etc/apache2/sites-available/$siteName-le-ssl.conf ]
    then
        echo -e "${Bold}${Green}SSL Successful for $siteName${Rst}"
        continue
    fi
    echo -e "${Bold}${Red}SSL unsuccessful for $siteName${Rst}, try again"
    exit
done

if [ $allSitesCount == -1 ]
then
    echo -e "${Red}SSL Installation exiting because one of the following were true"
    echo "1. You had no sites added (try running 'easy-apache -a' to add site"
    echo "2. You already have ssl certificates installed"
    echo -e "Please check & run SSL Installation again${Rst}"
    exit
fi

echo "Allowing 'Apache Full' in ufw"
sudo ufw delete allow 'Apache'
sudo ufw allow 'Apache Full'

echo "SSL added to sites:"
temp=-1
while [ $temp != $allSitesCount ]
do
    if [ -e /etc/apache2/sites-available/${allSitesURL[$temp]}-le-ssl.conf ]
    then
        temp=`expr $temp + 1`
        echo "https://${allSitesURL[$temp]}"
        continue
    fi
    temp=`expr $temp + 1`
    echo "http://"${allSitesURL[$temp]}
done

easy-apache.sh
#!/bin/bash

serverIP=`curl -s icanhazip.com`
banner() {
	curl https://raw.githubusercontent.com/realpvn/easy-apache/master/easy-banner.txt
}

terminalColors () {
	# Colors from - https://gist.github.com/5682077.git
	TC='\e['

	Bold="${TC}1m"    # Bold text only, keep colors
	Undr="${TC}4m"    # Underline text only, keep colors
	Rst="${TC}0m"     # Reset all coloring and style

	# Basic colors
	Black="${TC}30m"
	Red="${TC}31m"
	Green="${TC}32m"
	Yellow="${TC}33m"
	Blue="${TC}34m"
	Purple="${TC}35m"
	Cyan="${TC}36m"
	White="${TC}37m"

	Crossed="\u2718"

	#copy pasted tick symbold because unicode was acting weird
	Ticked="✓"
	Info="!"

	Oper=${Bold}'[ * ]'
	OperSuccess=${Bold}${Green}'[ '${Ticked}' ]'
	OperFailed=${Bold}${Red}'[ '${Crossed}' ]'
	Info=${Bold}${Yellow}'[ '${Info}' ]'
}

printNormal () {
	echo -e "${Oper} $1$Rst";
}

printSuccess() {
	echo -e "${OperSuccess} $1$Rst"
}

printFailed() {
	echo -e "${OperFailed} $1$Rst"
}

printInfo() {
	echo -e "${Info} $1$Rst"
}

help () {
	banner
	echo -e "${Bold}Usage:${Rst}\neasy-apache [options]\n\n${Bold}Options:${Rst}\n-f | --full:\t\tFull setup, default option if none is provided\n-a | --apache:\t\tAdding new site (includes apache install)\n-s | --SSL:\t\tInstall SSL certificate for sites available\n-h | --help:\t\tHelp (shows available commands)\n-v | --version:\t\tCheck easy-apache version${Rst}"
	echo -e "\n${Bold}Example\n${Rst}easy-apache -f\t\t\t#for full installation i.e Apache & SSL certificate\neasy-apache --full\t\t#for full installation i.e Apache & SSL certificate\neasy-apache -a\t\t\t#for installating Apache server\neasy-apache -s\t\t\t#for installating SSL certificate"
}

version() {
	version=`apt-cache show easy-apache | grep Version | cut -d ' ' -f2`
	echo $version
}

apacheInstall () {
	banner
	printInfo "Server Public IP: ${Yellow}${Bold}"${serverIP}${Rst}

	# Updating Server
	printNormal "Updating Server"
	sudo apt update -y &> /dev/null
	printNormal "Almost done, this might take few more mins"
	sudo apt upgrade -y &> /dev/null
	printSuccess "Updated"

	# Cleaning after server update
	sudo apt autoremove -y &> /dev/null
	sudo apt autoclean -y &> /dev/null
	printSuccess "Cleaned after server update"

	allSitesURL=""
	allSitesCount=-1

	# Installing apache2
	printSuccess "Apache Installed"
	printInfo "${Yellow}${Bold}Apache `apache2 -v`${Rst}"

	while true
	do
		printInfo "Setting up new site"
		read -p "Site URL (99 - exit): " siteURL

		if [[ -z "$siteURL" ]]
		then
			printFailed "Site URL empty"
			break
		fi

		if [ $siteURL == 99 ]
		then
			if [ $allSitesCount == -1 ]
			then
				echo "Good byee..."
				exit
			fi
			break
		fi

		re="(https:\/\/)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"
		if [[ ! $siteURL =~ $re ]]
		then
			printFailed "Invalid URL"
			continue
		fi

		printSuccess "URL Valid"

		if [ -e /etc/apache2/sites-available/$siteURL.conf ]
		then
			printFailed "$siteURL already exists"
			printInfo "Do you want to overwrite $siteURL? (Yy/Nn/99 to exit setup)"
			read overwriteSite
			case $overwriteSite in
				[Yy]* ) #allSitesURL used for printing at last
						allSitesCount=`expr $allSitesCount + 1`
						allSitesURL[$allSitesCount]=$siteURL
						addSite $siteURL
						break;;
				[Nn]* ) continue;;
				[99]* ) break;;
					* ) printInfo "Invalid input. Skipping $siteURL setup"
						continue;;
			esac
		else
			#allSitesURL used for printing at last
			allSitesCount=`expr $allSitesCount + 1`
			allSitesURL[$allSitesCount]=$siteURL
			addSite $siteURL
			break
		fi
	done

	if [ $allSitesCount == -1 ]
	then
		exit
	fi

	if sudo ufw status | grep -q inactive$
	then
		printInfo "UFW is disabled. You need to enable it to continue..."
		
		while true
		do
			read -p "Do you want to enable now (Yy/Nn)? " ufwEnable
			case $ufwEnable in
				[Yy]* ) echo "y" | sudo ufw enable &> /dev/null;
						printSuccess "UFW enabled"
						sudo ufw allow ssh &> /dev/null;
						sudo ufw allow Apache &> /dev/null;
						printSuccess "Allowed SSH & Apache in ufw"
						break;;
				[Nn]* ) printInfo "You cannot view the site until you enable ufw and allow SSH & Apache"; break;;
					* ) echo -e "Please answer yes(Yy) or no(Nn) ";;
			esac
		done
	else
		printSuccess "UFW already enabled"
	fi

	sudo a2dissite 000-default.conf &> /dev/null
	printSuccess "Disabled default site"

	sudo systemctl restart apache2
	printSuccess "Apache restart success"

	#TODO(pavank): try to check if site is available at server ${IP}

	printSuccess "Success! Your site(s) have been added successfully"
	printSuccess "Point your domains A record to $IP and after DNS propagation everything should be working fine.${Rst}"
	echo -e "Sites added and configured are:"

	temp=-1
	while [ $temp != $allSitesCount ]
	do
		if [ -e /etc/apache2/sites-available/${allSitesURL[$temp]}-le-ssl.conf ]
		then
			temp=`expr $temp + 1`
			echo -e "https://${allSitesURL[$temp]}"
			continue
		fi
		temp=`expr $temp + 1`
		echo -e "http://"${allSitesURL[$temp]}
	done
}

addSite () {
	siteURL=$1

	# Used for directory name (which is without domain TLD, example.com site folder would be "example" not "example.com")
	siteNameNoTLD=`echo -e $siteURL | cut -d'.' -f1`

	printInfo "Creating directory & adding permissions"
	sudo mkdir -p /var/www/$siteNameNoTLD
	sudo chown -R $USER:$USER /var/www/$siteNameNoTLD
	sudo chmod -R 755 /var/www
	printSuccess "Done"

	# Create temporary index.html page for viewing
	sudo echo -e "<h1>Server setup by <a href='https://github.com/realpvn/easy-apache.git'>easy-apache</a> (https://github.com/realpvn/easy-apache.git) </h1>" &> /var/www/$siteNameNoTLD/index.html

	printSuccess "Site $siteURL created"
	printNormal "Configuring"
	read -p "Email (leave blank if not required):" siteEmail
	if [ -z $siteEmail ]; then
		siteEmail=dev@localhost
	fi
	echo -e "<VirtualHost *:80>\n\tServerAdmin $siteEmail\n\tServerName $siteURL\n\tServerAlias www.$siteURL\n\tDocumentRoot /var/www/$siteNameNoTLD\n\tErrorLog \${APACHE_LOG_DIR}/error.log\n\tCustomLog \${APACHE_LOG_DIR}/access.log combined\n</VirtualHost>" | sudo tee /etc/apache2/sites-available/$siteURL.conf
	printSuccess "Site $siteURL configured successfully"

	sudo a2ensite $siteURL.conf &> /dev/null
	printSuccess "Enabled configuration for $siteURL"
}

sslInstall () {
	banner
	printNormal "SSL Installation"
	allSitesURL=""
	allSitesCount=-1

	# Install certbot to install and manage SSL
	sudo apt install certbot python3-certbot-apache -y &> /dev/null
	printSuccess "Installed Certbot"

	while true
	do
		read -p "Site URL to add SSL (99 - Exit): " siteName
		if [[ -z "$siteName" ]]
		then
			printFailed "Site URL empty"
			break
		fi

		if [ $siteName == 99 ]
		then
			if [ $allSitesCount == -1 ]
			then
				echo -e "Good byee..."
				exit
			fi
			break
		fi

		re="(https:\/\/)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"
		if [[ ! $siteName =~ $re ]]
		then
			printFailed "Invalid URL"
			printInfo "Example site URL - example.com"
			continue
		fi

		if [ ! -e /etc/apache2/sites-available/${siteName}.conf ]
		then
			printFailed "${siteName} does not exist"
			printInfo "Use 'easy-apache -a' to add a site"
			exit
		fi

		if [ -e /etc/apache2/sites-available/${siteName}-le-ssl.conf ]
		then
			printFailed "${siteName} already has SSL installed"
			continue
		fi

		allSitesCount=`expr $allSitesCount + 1`
		allSitesURL[$allSitesCount]=$siteName
		
		sudo certbot --apache -d www.$siteName -d $siteName
		if [ -e /etc/apache2/sites-available/$siteName-le-ssl.conf ]
		then
			printSuccess "SSL Successful for $siteName"
			continue
		fi
		printFailed "SSL unsuccessful for $siteName"
		printInfo "Make sure your domains A record is pointing to your IP: $serverIP${Rst}"
		exit
	done

	if [ $allSitesCount == -1 ]
	then
		printFailed "SSL Installation failed"
		printInfo "Reasons for failure"
		printInfo "1. You do not have any sites added (use 'easy-apache -a' to add site)"
		printInfo "2. You already have ssl certificates installed"
		printInfo "Please check & run SSL Installation again"
		exit
	fi

	sudo ufw delete allow 'Apache' &> /dev/null
	sudo ufw allow 'Apache Full' &> /dev/null
	printSuccess "Apache Full allowed in ufw"

	echo -e "SSL added to sites:"
	temp=-1
	while [ $temp != $allSitesCount ]
	do
		if [ -e /etc/apache2/sites-available/${allSitesURL[$temp]}-le-ssl.conf ]
		then
			temp=`expr $temp + 1`
			echo -e "https://${allSitesURL[$temp]}"
			continue
		fi
		temp=`expr $temp + 1`
		echo -e "http://"${allSitesURL[$temp]}
	done
}

terminalColors
if [ "$1" != "" ]
then
    PARAM=$1
    case $PARAM in
        -f | --full 	 ) apacheInstall; sslInstall; exit;;
        -a | --apache 	 ) apacheInstall; exit;;
        -s | --SSL       ) sslInstall; exit;;
        -h | --help      ) help; exit;;
		-v | --version   ) version; exit;;
                        *) help; exit;;
    esac
fi
help